src/content/en/updates/2020/11/puppetaria.md

```diff
@@ -0,0 +1,175 @@
+
+
+### DevTools test suite benchmark {: .benchmark }
+The previous benchmark showed that implementing our querying mechanism at the CDP layer gives a performance boost in a clinical unit-test scenario.
+
+To see if the difference is pronounced enough to make it noticeable in a more realistic scenario of running a full test suite, we [patched the DevTools end-to-end test suite](https://chromium-review.googlesource.com/c/devtools/devtools-frontend/+/2360013) to make use of the JavaScript and CDP-based prototypes and compared the runtimes. In this benchmark, we changed a total of 43 selectors from `[aria-label=…]` to a custom query handler `aria/…`, which we then implemented using each of the prototypes.
+
+Some of the selectors are used multiple times in test scripts, so the actual number of executions of the `aria` query handler was 113 per run of the suite. The total number of query selections was 2253, so only a fraction of the query selections happened through the prototypes.
+
+
+![Benchmark: e2e test suite](/web/updates/images/2020/11/puppetaria/puppetaria-03.svg)
+
+As seen in the figure above, there is a discernible difference in the total runtime. The data is too noisy to conclude anything specific, but it is clear that the performance gap between the two prototypes shows in this scenario as well.
+### A new CDP endpoint
+In light of the above benchmarks, and since the launch flag-based approach was undesirable in general, we decided to move forward with implementing a new CDP command for querying the accessibility tree. Now, we had to figure out the interface of this new endpoint.
+
+For our use case in Puppeteer, we need the endpoint to take so-called [`RemoteObjectIds`](https://chromedevtools.github.io/devtools-protocol/tot/Runtime/#type-RemoteObjectId) as argument and, to enable us to find the corresponding DOM elements afterwards, it should return a list of objects that contains the [`backendNodeIds`](https://chromedevtools.github.io/devtools-protocol/tot/DOM/#type-BackendNodeId) for the DOM elements. 
+
+As seen in the chart below, we tried quite a few approaches satisfying this interface. From this, we found that the size of the returned objects, i.e whether or not we returned full accessibility nodes or only the `backendNodeIds` made no discernible difference. On the other hand, we found that using the existing [`NextInPreOrderIncludingIgnored`](codesearch link) was a poor choice for implementing the traversal logic here, as that yielded a noticeable slow-down.
```

Comment:
@mathiasbynens Can you confirm if this is the right link? https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/modules/accessibility/ax_object.cc;l=3226;bpv=1;bpt=1?q=NextInPreOrderIncludingIgnored

Reply:
Opps, did not highlight the text - for the line mentioned `(codesearch link)`.

Reply:
The link from the doc was `https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/modules/accessibility/ax_object.h;l=977-982;drc=92569ee6b4ff7e0344f7780d2b2e843b1c359d05`